src/Workspaces/Core/Portable/Utilities/TaskQueue.cs

```diff
@@ -0,0 +1,102 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Shared.TestHooks;
+
+namespace Roslyn.Utilities
+{
+    internal sealed class TaskQueue
+    {
+        private readonly IAsynchronousOperationListener _operationListener;
+        private readonly TaskScheduler _taskScheduler;
+
+        private readonly object _gate = new object();
+        private Task _latestTask;
+
+        internal TaskQueue(IAsynchronousOperationListener operationListener, TaskScheduler taskScheduler)
+        {
+            Contract.ThrowIfNull(operationListener);
+            Contract.ThrowIfNull(taskScheduler);
+
+            _operationListener = operationListener;
+            _taskScheduler = taskScheduler;
+            _latestTask = Task.CompletedTask;
+        }
+
+        public Task LastScheduledTask => _latestTask;
+
+        private IAsyncToken BeginOperation(string taskName)
+            => _operationListener.BeginAsyncOperation(taskName);
+
+        private TTask EndOperation<TTask>(IAsyncToken token, TTask task) where TTask : Task
+        {
+            _ = task.CompletesAsyncOperation(token);
+            return task;
```

Comment:
up to you if  you want to doc this.  might be nice for future readers.

Reply:
Why do you think is worth documenting here?